File: zig_renderer/src/Texture.zig
const std = @import("std");
const Self = @This();
const gl = @import("gl");

id: c_uint = undefined,

pub fn init(width: u32, height: u32, grayscale: []const u8) Self {
    var self = Self{};
    gl.genTextures(1, &self.id);
    gl.bindTexture(gl.GL_TEXTURE_2D, self.id);
    gl.texImage2D(
        gl.GL_TEXTURE_2D,
        0,
        gl.GL_ALPHA,
        @intCast(c_int, width),
        @intCast(c_int, height),
        0,
        gl.GL_ALPHA,
        gl.GL_UNSIGNED_BYTE,
        &grayscale[0],
    );
    gl.texParameteri(gl.GL_TEXTURE_2D, gl.GL_TEXTURE_MIN_FILTER, gl.GL_NEAREST);
    gl.texParameteri(gl.GL_TEXTURE_2D, gl.GL_TEXTURE_MAG_FILTER, gl.GL_NEAREST);
    // std.debug.assert(gl.getError() == 0);
    return self;
}

pub fn deinit(self: Self) void {
    gl.deleteTextures(1, &self.id);
}

pub fn bind(self: Self) void {
    gl.bindTexture(gl.GL_TEXTURE_2D, self.id);
}
pub fn unbind(_: Self) void {
    gl.bindTexture(gl.GL_TEXTURE_2D, 0);
}

File: cpp/gl_renderer/texture.cpp
#include "texture.h"
#include <assert.h>
#include <glad/gl.h>

//
// Texture
//
Texture::Texture() { glGenTextures(1, &_id); }
Texture::~Texture() { glDeleteTextures(1, &_id); }
void Texture::bind() { glBindTexture(GL_TEXTURE_2D, _id); }
void Texture::unbind() { glBindTexture(GL_TEXTURE_2D, 0); }

std::shared_ptr<Texture> Texture::load(int w, int h,
                                       const unsigned char *data) {
  auto texture = std::make_shared<Texture>();
  texture->bind();
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
  glTexImage2D(GL_TEXTURE_2D, 0, GL_ALPHA, w, h, 0, GL_ALPHA, GL_UNSIGNED_BYTE,
               data);
  texture->unbind();
  return texture;
}
